RiotSwiftUI/Modules/Template/TemplateAdvancedRoomsExample/TemplateRoomChat/Service/MatrixSDK/TemplateRoomChatService.swift

```diff
@@ -0,0 +1,134 @@
+// 
+// Copyright 2021 New Vector Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+import Combine
+
+@available(iOS 14.0, *)
+class TemplateRoomChatService: TemplateRoomChatServiceProtocol {
+    
+    // MARK: - Properties
+    
+    // MARK: Private
+    
+    private let room: MXRoom
+    private let eventFormatter: EventFormatter
+    private var roomState: MXRoomState?
+    private var roomListenerReference: Any?
+    
+    // MARK: Public
+    private(set) var chatMessagesSubject: CurrentValueSubject<[TemplateRoomChatMessage], Never>
+    private(set) var roomInitializationStatus: CurrentValueSubject<TemplateRoomChatRoomInitializationStatus, Never>
+    
+    var roomName: String? {
+        self.room.summary.displayname
+    }
+    // MARK: - Setup
+    
+    init(room: MXRoom) {
+        self.room = room
+        self.eventFormatter = EventFormatter(matrixSession: room.mxSession)
+        self.chatMessagesSubject = CurrentValueSubject([])
+        self.roomInitializationStatus = CurrentValueSubject(.notInitialized)
+        
+        initializeRoom()
+    }
+    
+    deinit {
+        guard let reference = roomListenerReference else { return }
+        room.removeListener(reference)
+    }
+    
+    // MARK: Public
+    func send(textMessage: String) {
+        var localEcho: MXEvent? = nil
+        room.sendTextMessage(textMessage, localEcho: &localEcho, completion: { _ in })
+    }
+    
+    // MARK: Private
+    
+    private func initializeRoom(){
+        room.state { [weak self] roomState in
+            guard let self = self else { return }
+            if let roomState = roomState {
+                self.roomState = roomState
+                self.roomInitializationStatus.value = .initialized
+                self.loadInitialMessages()
+                self.startListeningToRoomEvents()
+            } else {
+                self.roomInitializationStatus.value = .failedToInitialize
+            }
+        }
+    }
+    
+    private func loadInitialMessages() {
+        let batch = room.enumeratorForStoredMessages.nextEventsBatch(200)
```

Comment:
You should use `MXRoom.liveTimeline` to get messages. It will fetch messages from the server if not enough in the store

You should listen to it to get new messages.
Its loading is async but you get the room state for it. You can load it from  initializeRoom

Reply:
Great thanks for that.